subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -91,6 +172,812 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+	}
+
+	NET_INFO("Response status %s", rsp->http_status);
+
+	*status = rsp->http_status_code;
+}
+
+static int exec_http_query(struct net_if *iface, int sock)
+{
+	int32_t timeout = HTTP_WAIT_TIME;
+	struct http_request req;
+	uint16_t status;
+	int ret;
+
+	memset(&req, 0, sizeof(req));
+
+	req.method = HTTP_GET;
+	req.url = online_check->url;
+	req.host = online_check->host;
+	req.port = online_check->port;
+	req.protocol = "HTTP/1.1";
+	req.response = response_cb;
+	req.recv_buf = online_check->http_recv_buf;
+	req.recv_buf_len = sizeof(online_check->http_recv_buf);
+
+	/* Make sure that we cannot access recv_buf data multiple times */
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+	ret = http_client_req(sock, &req, timeout, &status);
+	if (ret < 0) {
+		NET_DBG("HTTP request failed (%d)", ret);
+		goto out;
+	}
+
+	/* Lets assume online also for 301 "Moved Permanently" */
+	if (status == 200 || status == 301) {
+		NET_DBG("Sending Online Connectivity event for interface %d",
+			net_if_get_by_iface(iface));
+		net_mgmt_event_notify(NET_EVENT_CONNECTIVITY_ONLINE, iface);
+
+		enable_online_state_verifier();
+	} else {
+		NET_DBG("Received HTTP status %d, not considering online.",
+			status);
+	}
+
+out:
+	k_mutex_unlock(&conn_mgr_mon_lock);
+
+	return ret;
+}
+
+static int do_online_http_check(struct net_if *iface, const char *host)
+{
+	int proto = IPPROTO_TCP;
+	const sec_tag_t *tags = NULL;
+	size_t tags_size = 0;
+	socklen_t addrlen;
+	int sock;
+	int ret;
+
+	if (!online_check->hostaddr_valid) {
+		ret = resolve_url(host);
+		if (ret < 0) {
+			NET_DBG("Cannot parse URL \"%s\" (%d)",
+				host, ret);
+			return ret;
+		}
+	}
+
+	if (!online_check->hostaddr_valid) {
+		return -EAGAIN;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS) && online_check->is_tls) {
+		if (online_check->cb == NULL) {
+			NET_DBG("HTTPS request but callback not registered. "
+				"HTTP online check disabled.");
+			return -ENOTSUP;
+		}
+
+		ret = online_check->cb(iface, &tags, &tags_size,
+				       host,
+				       online_check->host,
+				       online_check->port,
+				       &online_check->hostaddr,
+				       online_check->user_data);
+		if (ret < 0) {
+			NET_DBG("Setting up socket failed (%d)", ret);
+			return ret;
+		}
+
+		if (tags_size == 0 || tags == NULL) {
+			NET_DBG("Security tags is empty, TLS disabled");
+		} else {
+			proto = IPPROTO_TLS_1_2;
+		}
+	}
+
+	sock = zsock_socket(online_check->hostaddr.sa_family,
+			    SOCK_STREAM, proto);
+	if (sock < 0) {
+		NET_DBG("Socket creation failed (%d)", sock);
+		return ret;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS) && proto == IPPROTO_TLS_1_2) {
+		ret = zsock_setsockopt(sock, SOL_TLS, TLS_SEC_TAG_LIST,
+				       tags, tags_size);
+		if (ret < 0) {
+			LOG_ERR("TLS setsockopt: %d", errno);
+			ret = -errno;
+			goto out;
+		}
+
+#define TLS_PEER_HOSTNAME "localhost"
+		ret = zsock_setsockopt(sock, SOL_TLS, TLS_HOSTNAME,
+				       TLS_PEER_HOSTNAME,
```

Comment:
This shouldn't be hardcoded, we should set the TLS hostname to the actual value the certificate was issued for (could be taken from the URL I think, or provided with the callback that feeds certs). Not setting the hostname properly will result in TLS handshake failure with real servers.

Reply:
Indeed, good point. I will figure out something for the hostname.